### PR TITLE
refactor(x/swaprouter, x/gamm): unify swap routes in swaprouter

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #3808

## What is the purpose of the change

This PR removes the swap routes in `x/gamm` and refactors all legacy `x/gamm` swap messages and methods to utilize the new swap route structs from the `x/swaprouter` module.

This is done to reduce code duplication and redundant conversions between the different swap routes.

## Brief Changelog
- removed proto declarations for all `x/gamm `swap routes but [`SwapExactAmountInRoute`](https://github.com/osmosis-labs/osmosis/blob/a3afcda5ca7694ea40a4fa573db19ffdae2d0a3b/proto/osmosis/gamm/v1beta1/tx.proto#L86-L90)
   * It is kept for the serialization / deserialization test but is to be removed in the future.
- substituted usage of old `x/gamm` swap routes with the `x/swaprouter` swap routes.

## Testing and Verifying

This change added tests and can be verified as follows:
https://github.com/osmosis-labs/osmosis/blob/a3afcda5ca7694ea40a4fa573db19ffdae2d0a3b/x/gamm/types/route_test.go#L16

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? yes
  - How is the feature or change documented? not applicable

### CODE REVIEW

**Comment by p0mvn** [comment]
While this is not state-breaking, I added the label still to be more careful with this PR as it has an important API break.

File: proto/osmosis/gamm/v1beta1/tx.proto [hunk 1]
```diff
@@ -3,16 +3,17 @@ package osmosis.gamm.v1beta1;
...
 
 service Msg {
   rpc JoinPool(MsgJoinPool) returns (MsgJoinPoolResponse);
   rpc ExitPool(MsgExitPool) returns (MsgExitPoolResponse);
   rpc SwapExactAmountIn(MsgSwapExactAmountIn)
-      returns (MsgSwapExactAmountInResponse);
+      returns (MsgSwapExactAmountInResponse) {}
   rpc SwapExactAmountOut(MsgSwapExactAmountOut)
-      returns (MsgSwapExactAmountOutResponse);
+      returns (MsgSwapExactAmountOutResponse) {}
```
Comment (czarcas7ic) [inline proto/osmosis/gamm/v1beta1/tx.proto:16]: Why this change?

**Review by czarcas7ic — APPROVED** [review]

File: proto/osmosis/gamm/v1beta1/tx.proto (hunk 1 above)
Comment (p0mvn) [reply proto/osmosis/gamm/v1beta1/tx.proto:16]: was trying to deprecate those and failed to revert properly. Going to fix

File: x/gamm/types/route_test.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	swapRouterExactAmountInRoute := swaproutertypes.SwapAmountInRoute{}
+	gammExactAmountInRoute := gammtypes.SwapAmountInRoute{}
+
+	swapRouterBz, err := proto.Marshal(&swapRouterExactAmountInRoute)
+	require.NoError(t, err)
+	gammBz, err := proto.Marshal(&gammExactAmountInRoute)
+	require.NoError(t, err)
+
+	require.Equal(t, swapRouterBz, gammBz)
+}
```
Comment (mattverse) [inline x/gamm/types/route_test.go:26]: What's the reason we do not delete `SwapAmountInRoute` from gamm types completely? Seems that `SwapAmountInRoute` in gamm types have no actual usage and its not getting called anywhere

**Review by mattverse — COMMENTED** [review]
Left one question / concerning point about gamm types. Other then that, looks LGTM

File: x/gamm/types/route_test.go (hunk 2 above)
Comment (p0mvn) [reply x/gamm/types/route_test.go:26]: We can delete that in a subsequent PR. It is currently being used as proof that there should be no breakage

File: proto/osmosis/gamm/v1beta1/query.proto [hunk 3]
```diff
@@ -282,7 +283,7 @@ message QuerySwapExactAmountOutRequest {
   // TODO: CHANGE THIS TO RESERVED IN A PATCH RELEASE
   string sender = 1 [ (gogoproto.moretags) = "yaml:\"sender\"" ];
   uint64 pool_id = 2 [ (gogoproto.moretags) = "yaml:\"pool_id\"" ];
-  repeated SwapAmountOutRoute routes = 3 [
+  repeated osmosis.swaprouter.v1beta1.SwapAmountOutRoute routes = 3 [
     (gogoproto.moretags) = "yaml:\"routes\"",
     (gogoproto.nullable) = false
   ];
```
Comment (ValarDragon) [inline proto/osmosis/gamm/v1beta1/query.proto:289]: We should ask dan lynch and front-end how breaking this is.

Its not clear to me that its worth breaking this in the gamm proto, vs just doing a conversion in the proto boilerplate handling code.

Comment (ValarDragon) [reply proto/osmosis/gamm/v1beta1/query.proto:289]: (I think it will break their compiled types, unless we can register a way for a type to get aliased or something in codegen)

File: x/gamm/types/route_test.go (hunk 2 above)
Comment (ValarDragon) [reply x/gamm/types/route_test.go:26]: This doesn't guarantee things, we need non-empty entries for every field for this to actually be checked.

Comment (p0mvn) [reply x/gamm/types/route_test.go]: Missed adding values. Added now

Comment (mattverse) [reply x/gamm/types/route_test.go]: https://github.com/osmosis-labs/osmosis/issues/3921

File: proto/osmosis/gamm/v1beta1/query.proto (hunk 3 above)
Comment (pyramation) [reply proto/osmosis/gamm/v1beta1/query.proto:287]: as long as the interfaces/API don't change, it doesn't matter where the types live. Telescope will handle it.

from [what I can see here](https://github.com/osmosis-labs/osmosis/blob/f594d0d6614a46f4cbf5dfb6af71ace3828b41d9/proto/osmosis/swaprouter/v1beta1/swap_route.proto#L8-L18) it looks the the interface to the type doesn't change, so will be totally ok!

Comment (pyramation) [reply proto/osmosis/gamm/v1beta1/query.proto:287]: also, even w/o telescope, from what I can tell (since I believe osmosis-fe uses another tool) — nothing should break. But, we should probably all keep in sync when this update rolls out to update FE libs and repos.

Comment (p0mvn) [reply proto/osmosis/gamm/v1beta1/query.proto:287]: Thanks, @pyramation . I also spoke with Jon to make sure that there is no unexpected breakage for FE.

We will track this change and remind everyone to update before this is out.

This should be good to go

**Review by ValarDragon — APPROVED** [review]

**Comment by ValarDragon** [comment]
Needs changelog entry for breaking changes, otherwise LGTM.

**Comment by p0mvn** [comment]
Seems to be JSON serialization compatible based on added test here: [3a46711](https://github.com/osmosis-labs/osmosis/pull/3880/commits/3a4671110656937ccb3cc26cd8d8b5cbe44aac97)

**Comment by ValarDragon** [comment]
oh awesome, thats great!
